spec/index.bs

```diff
@@ -1338,20 +1344,112 @@ Only supported capabilities may be requested in {{GPUAdapter/requestDevice()}};
 requesting unsupported capabilities results in failure.
 
 The capabilities of a [=device=] are determined in "[=a new device=]" by starting with the adapter's
-defaults (no features and the default [=supported limits=])
+defaults (from {{adapter/[[default feature level]]}})
 and adding capabilities as requested in {{GPUAdapter/requestDevice()}}.
 These capabilities are enforced regardless of the capabilities of the [=adapter=].
 
 <p tracking-vector>
 For privacy considerations, see [[#privacy-machine-limits]].
 
+### Feature Levels ### {#feature-levels}
+
+<dfn enum>GPUFeatureLevel</dfn>s describe a bundles of [=features=] and [=limits=].
+They are passed to {{GPU/requestAdapter()}}
+(via {{GPURequestAdapterOptions}}.{{GPURequestAdapterOptions/minFeatureLevel}}),
+which return either `null` or an adapter which:
+
+- Provides certain additional capabilities by default when requesting a device.
+- May have additional adapter capability guarantees.
+
+Note:
+Feature levels are defined additively on top of previous feature levels, such that it is always
+possible to request a lower feature level, then choose additional capabilities from there to
+create a device that could have been created from a higher feature level.
+This avoids the need to call {{GPU/requestAdapter()}} multiple times to initialize an application:
+request the minimum {{GPURequestAdapterOptions/minFeatureLevel}} needed for base functionality,
+then apply progressive enhancements depending on what extra capabilities are available on the
+resulting adapter.
+
+{{GPU/requestAdapter()}} may fail (if the implementation or underlying implementation does
+not support the feature level), but it must gracefully accept all enum values defined here.
+If any feature level is requested on an implementation that only supports lower feature levels,
+the implementation must return `null`.
+
+Implementations must not automatically upgrade feature level requests, unless the requested
+feature level is <dfn dfn for=GPUFeatureLevel>upgradeable</dfn> to another feature level.
+<!-- This is defined on a level-by-level basis. For example, someday, FL1 could allow upgrading to
+FL2, if we want to allow implementation to drop their implementations of FL1 validation entirely. -->
```

Comment:
I was imagining something like this, though of course we wouldn't have to figure out the actual DAG until later 
![image](https://github.com/user-attachments/assets/26dc1942-9fa6-4f26-9c70-6ef465309bd7)

Your suggestion in pink dashed line - I think in practice it would only happen if an implementation dropped fl1-desktop AND didn't have fl2-desktop or fl3-desktop, but did have fl3-mobile.

Reply:
Giving this some additional thought: I think that the fact that this says (in the commented out text) "This is defined on a level-by-level basis" is part of what has given me pause here. I feel like it would be better if we could give a definitive, all encompassing definition of upgradeable here. For example, if we define each feature level solely in terms of:
 - An array of required features
 - A set of limits

Then we can define upgradable as any feature level which is a superset of the same features and for which all limits are equal or better. Unfortunately compat doesn't fit nicely into that model unless we have a "core" feature, which then validates the concern voiced by others that we'd be effectively redefining compat as the base feature level. It also doesn't nicely allow for things like "fl1 requires BC texture compression OR ETC2" but honestly I feel like those types of features should probably be left out of feature levels altogether (unless we have specifically mobile/desktop focused feature levels)

Also, I don't think I appreciated previously that the feature level of the adapter you get back isn't actually reported anywhere, which I think is a good thing. It means that if you request "fl1" and get an adapter back at all, you know that it will have those baseline features/limits without worrying too much about what else might get added on top. As long as all of the features are reported in the device.features list, regardless of if they're part of the feature level or not, then the upgrades can be more-or-less silent, with the only way to detect it being to scrub through the feature list and compare against known feature levels. But I can't think of any motivation to do that.

Reply:
> Unfortunately compat doesn't fit nicely into that model unless we have a "core" feature, which then validates the concern voiced by others that we'd be effectively redefining compat as the base feature level.

I think we _should_ have this - more specifically, a list of features that get compat back up to core - see `'everything-from-compat-to-fl1'` in the [prose version of the proposal](https://github.com/gpuweb/gpuweb/issues/4656#issuecomment-2121487301).

> Then we can define upgradable as any feature level which is a superset of the same features and for which all limits are equal or better.

I wanted the spec to be very strict/stingy about when browsers are allowed to do upgrades. Compat would be upgradeable to fl1 from the start, of course, but when we add fl2, fl1 would _not_ be upgradeable to fl2 initially. We wouldn't say fl1 is upgradeable to fl2 until the WG makes the decision that browsers are allowed to start dropping their fl1 implementations (if ever).